services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/AttributeName.java

```diff
@@ -0,0 +1,140 @@
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+/**
+ * High-level representation of a DynamoDB 'attribute name' that can be used in various situations where the API requires
+ * or accepts an Attrbute Name. In addition various convenience methods are provided to help manipulate expressions.
+ * <p>
+ * At a minimum, an AttributeName must contain a string that represents the name of the Attribute.
+ * <p>
+ * Optionally, nested attribute names can be added with tokens using the addNestedAttributeName .
+ * Every call to addNestedAttributeName represent next level of Nesting Attribute.
+ * If we want to denote three level of Nesting attribute name then addNestedAttributeName should be called thrice.
+ *
+ * <p>
+ * Example, if we want to represent a two level nested attribute outerAttribute.innerAttributeLevel1.innerAttributeLevel2
+ * then it can be represented as :-
+ * {@code
+ * AttributeName myAttributeName = AttributeName.builder()
+ * .name("outerAttribute).addNested("innerAttributeLevel1).nestedAttributeName("innerAttributeLevel2").build();
+ * }
+ */
+@SdkPublicApi
+public final class AttributeName {
+
+    private final String name;
+    private final List<String> nestedAttributeNames;
+
+    private AttributeName(String name, List<String> nestedAttributeNames) {
+        this.name = name;
+        this.nestedAttributeNames = nestedAttributeNames != null
+                ? Collections.unmodifiableList(nestedAttributeNames) : null;
+    }
+
+    /**
+     * Construct a new AttributeNameBuilder.
+     *
+     * @return a new AttributeNameBuilder.
+     */
+    public static AttributeNameBuilder builder() {
+        return new Builder();
+    }
+
+    /**
+     * @return Name of the attribute.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * @return List of nested attributes , each entry in the list represent one level of nesting.
+     */
+    public List getNestedAttributeNames() {
+        return nestedAttributeNames;
+    }
+
+    /**
+     * First Builder Step in charge of setting Attribute Name.
+     * Next Builder available : addNestedAttribute.
+     */
+    public interface AttributeNameBuilder {
+        BuildStep name(String attributeToProject);
+    }
+
+    /**
+     * This is the final step in charge of building the AttributeName Object.
+     */
+    public interface BuildStep {
+        BuildStep addNested(String nestedAttributeName);
+
+        AttributeName build();
+    }
+
+    /**
+     * A builder for {@link AttributeName}
+     * This builder follows BuilderStep pattern.
+     * The Attribute Name always comes first followed by optional Nested Attribute Name.
+     */
+    public static class Builder implements AttributeNameBuilder, BuildStep {
+        private final List<String> nestedAttributeNames = new ArrayList<>();
```

Comment:
Our pattern is to initialize this to null like other parameters, and when they add something check if it's null and create a new ArrayList then if necessary. This also allows us to provide a setter method, although it's important when we let them set the whole list we always create our own new ArrayList to avoid mutation of the list outside of our control after object created : 

```
public Builder nestedAttributeNames(List<String> nestedAttributeNames) { ... }
```